src/react/reconcilation.js

```diff
@@ -107,29 +160,50 @@ export class Reconciler {
   ) {
     let value;
     let childContext = context;
+
     if (valueIsClassComponent(this.realm, componentType)) {
-      if (branchStatus !== "ROOT") {
-        throw new ExpectedBailOut("only class components at the root of __registerReactComponentRoot() are supported");
+      // We first need to know what type of class component we're dealing with.
+      // A "simple" class component is defined as:
+      //
+      // - having only a "render" method or many method, i.e. render(), _renderHeader(), _renderFooter()
+      // - having no lifecycle events
+      // - having no state
+      // - having no instance variables
+      //
+      // the only things a class component should be able to access on "this" are:
+      // - this.props
+      // - this.context
+      // - this._someRenderMethodX() etc
+      //
+      // Otherwise, the class component is a "complex" one.
+      // To begin with, we don't know what type of component it is, so we try and render it as if it were
+      // a simple component using the above heuristics. If an error occurs during this process, we assume
+      // that the class wasn't simple, then try again with the "complex" heuristics.
+      if (componentType.$IsSimpleReactClassComponent === undefined) {
+        try {
+          value = this._renderSimpleClassComponent(componentType, props, context, branchStatus, branchState);
+          componentType.$IsSimpleReactClassComponent = true;
+        } catch (e) {
+          // mark the component as not simple
+          componentType.$IsSimpleReactClassComponent = false;
+        }
+      } else if (componentType.$IsSimpleReactClassComponent === true) {
+        value = this._renderSimpleClassComponent(componentType, props, context, branchStatus, branchState);
```

Comment:
So if I understand this correctly, we assume that it is a class component if it passes the first time but if it fails the second time (e.g. if the known props are different for the second usage of the class), then we assume that it is a legit error?

Seems a little random and fragile since just reordering the order that these two cases get serialized will either fail or pass.

Should we just always try the simple path if it is possible for every particular render?

Reply:
I've changed how this works slightly so it's clearer. It now looks for an explicit `SimpleClassBailOut` error thrown to determine if the class component is indeed simple. If it is not simple, it will always from there on apply the "complex" route for that component. The `SimpleClassBailOut` cannot be thrown from normal bailouts so it's safe.